protocols/autonat/src/behaviour.rs

```diff
@@ -351,77 +348,155 @@ impl NetworkBehaviour for Behaviour {
         } else {
-            let connections = self.connected.get_mut(peer).expect("Peer is connected.");
-            connections.remove(conn);
+            let connections = self
+                .connected
+                .get_mut(&peer_id)
+                .expect("Peer is connected.");
+            connections.remove(&connection_id);
         }
     }
 
-    fn inject_dial_failure(
+    fn on_dial_failure(
         &mut self,
-        peer: Option<PeerId>,
-        handler: Self::ConnectionHandler,
+        peer_id: Option<PeerId>,
+        handler: <RequestResponse<AutoNatCodec> as NetworkBehaviour>::ConnectionHandler,
         error: &DialError,
     ) {
-        self.inner.inject_dial_failure(peer, handler, error);
-        if let Some(event) = self.as_server().on_outbound_dial_error(peer, error) {
+        self.inner.on_swarm_event(FromSwarm::DialFailure {
+            peer_id,
+            handler,
+            error,
+        });
+        if let Some(event) = self.as_server().on_outbound_dial_error(peer_id, error) {
             self.pending_out_events
                 .push_back(Event::InboundProbe(event));
         }
     }
 
-    fn inject_address_change(
+    fn on_address_change(
         &mut self,
-        peer: &PeerId,
-        conn: &ConnectionId,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
         old: &ConnectedPoint,
         new: &ConnectedPoint,
     ) {
-        self.inner.inject_address_change(peer, conn, old, new);
+        self.inner.on_swarm_event(FromSwarm::AddressChange {
+            peer_id,
+            connection_id,
+            old,
+            new,
+        });
 
         if old.is_relayed() && new.is_relayed() {
             return;
         }
-        let connections = self.connected.get_mut(peer).expect("Peer is connected.");
+        let connections = self
+            .connected
+            .get_mut(&peer_id)
+            .expect("Peer is connected.");
         let addr = new.get_remote_address();
         let observed_addr =
             if !new.is_relayed() && (!self.config.only_global_ips || addr.is_global_ip()) {
                 Some(addr.clone())
             } else {
                 None
             };
-        connections.insert(*conn, observed_addr);
+        connections.insert(connection_id, observed_addr);
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = <RequestResponse<AutoNatCodec> as NetworkBehaviour>::ConnectionHandler;
+    type OutEvent = Event;
+
+    fn inject_connection_established(
+        &mut self,
+        peer: &PeerId,
+        conn: &ConnectionId,
+        endpoint: &ConnectedPoint,
+        failed_addresses: Option<&Vec<Multiaddr>>,
+        other_established: usize,
+    ) {
+        let failed_addresses = failed_addresses
+            .map(|v| v.as_slice())
+            .unwrap_or_else(|| &[]);
+        self.on_connection_established(*peer, *conn, endpoint, failed_addresses, other_established)
+    }
```

Comment:
`Swarm` still calls the old `NetworkBehaviour::inject_*` methods. The `NetworkBehaviour::inject_*` method default implementations delegate to the `NetworkBehaviour::on_*` methods.

I suggest that this pull request ports `libp2p-autonat` to implement `NetworkBehaviour::on_*` only, and no more implement the `NetworkBehaviour::inject_*` methods.  For now `Swarm` will be calling `NetworkBehaviour::inject_*` whose default implementations would then call the `libp2p-autonat` `<Behaviour as NetworkBehaviour>::on_event` methods. Thus there is no need for `Behaviour` to implement `NetworkBehaviour::inject_*`.

Long term we want to (a) get rid of the `NetworkBehaviour::inject_*` methods and (b) call the `NetworkBehaviour::on_*` methods in `Swarm` directly.

Does the above make sense @jxs?

Reply:
yeah ofc! :) missed that, thanks Max. updated